### PR TITLE
Features/snpeff recipe

### PR DESCRIPTION
**Pull request by northwestwitch**
Moved code from mip to snpeff recipe (splitted in analysis and analysis_rio) . Also fixed a bug that prevented Conda.pm recipe to work. Tested on rasta and seems to work!

### CODE REVIEW

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 1]
```diff
@@ -0,0 +1,1256 @@
...
+Readonly my $ASTERISK   => q{*};
+Readonly my $DOT        => q{.};
+Readonly my $EMPTY_STR  => q{};
+Readonly my $NEWLINE    => qq{\n};
+Readonly my $SPACE      => q{ };
+Readonly my $UNDERSCORE => q{_};
+
+sub analysis_snpeff {
+
+## Function : snpeff annotates variants from different sources.
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: snpeff -> Snpeff

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 2]
```diff
@@ -0,0 +1,1256 @@
...
+## Returns  : |$xargs_file_counter
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Remove |$xargs_file_counter

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 3]
```diff
@@ -0,0 +1,1256 @@
...
+    my $infile_prefix  = $family_id . $infile_tag . $call_type;
+    my $outfile_prefix = $family_id . $outfile_tag . $call_type;
+
+    ## Paths
+    my $file_path_prefix    = catfile( $temp_directory, $infile_prefix );
+    my $outfile_path_prefix = catfile( $temp_directory, $outfile_prefix );
+
+    ### Assign suffix
+    ## Return the current infile vcf compression suffix for this jobid chain
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Remove empty line

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 4]
```diff
@@ -0,0 +1,1256 @@
...
+      qw{ slurm_submit_job_sample_id_dependency_add_to_family };
+    use MIP::Program::Variantcalling::Mip_vcfparser qw{ mip_vcfparser };
+    use MIP::Program::Variantcalling::Snpeff qw{ snpeff_ann };
+    use MIP::Program::Variantcalling::Snpsift
+      qw{ snpsift_annotate snpsift_dbnsfp };
+    use MIP::QC::Record qw{ add_program_outfile_to_sample_info };
+    use MIP::Recipes::Analysis::Xargs qw{ xargs_command };
+    use MIP::Script::Setup_script qw{ setup_script };
+    use MIP::Set::File qw{ set_file_suffix };
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: ## Constant
    Readonly my $VCFPARSER_OUTFILE_COUNT =>
$active_parameter_href->{vcfparser_outfile_count} - 1;

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 5]
```diff
@@ -0,0 +1,1256 @@
...
+            suffix_key     => q{variant_file_suffix},
+        }
+    );
+
+    my $vcfparser_analysis_type = $EMPTY_STR;
+
+    # Set default
+    my $vcfparser_contigs_ref = \@{ $file_info_href->{contigs_size_ordered} };
+
+    for (
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Replace with:
    ## Determined by vcfparser output
for my $vcfparser_outfile_counter ( 0 .. $VCFPARSER_OUTFILE_COUNT ) {

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 6]
```diff
@@ -0,0 +1,1256 @@
...
+                    java_use_large_pages =>
+                      $active_parameter_href->{java_use_large_pages},
+                    memory_allocation  => q{Xmx4g -XX:-UseConcMarkSweepGC},
+                    program_info_path  => $program_info_path,
+                    temp_directory     => $temp_directory,
+                    XARGSFILEHANDLE    => $XARGSFILEHANDLE,
+                    xargs_file_counter => $xargs_file_counter,
+                }
+            );
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 7]
```diff
@@ -0,0 +1,1256 @@
...
+            foreach my $contig ( @{$vcfparser_contigs_ref} ) {
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: my $snpeff_file_path_prefix = $file_path_prefix
+                          . $UNDERSCORE
+                          . $contig
+                          . $vcfparser_analysis_type;

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 8]
```diff
@@ -0,0 +1,1256 @@
...
+                          . q{stderr.txt},
+                        stdoutfile_path => $file_path_prefix
+                          . $UNDERSCORE
+                          . $contig
+                          . $vcfparser_analysis_type
+                          . $infile_suffix
+                          . $DOT
+                          . $xargs_file_counter,
+                        verbosity => q{v},
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Remove empty line

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 9]
```diff
@@ -0,0 +1,1256 @@
...
+        ## SnpSift Annotation
+        say {$FILEHANDLE} q{## SnpSift Annotation};
+
+        my $annotation_file_counter = 0;
+        my $xargs_file_path_prefix;
+
+        # Annotate using snpeff
+        if ( $active_parameter_href->{snpeff_ann} == 1 ) {
+            ## Create file commands for xargs
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: my $java_snpeff_jar = catfile(
+                        $active_parameter_href->{snpeff_path},
+                        q{snpEff.jar}
+                    );

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 10]
```diff
@@ -0,0 +1,1256 @@
...
+                }
+            );
+            ## Get parameters
+            my $name_prefix;
+            my $info_key;
+            if ( defined $annotation_info_key ) {
+
+                ## Apply specific INFO field output key for easier downstream processing
+                if (
+                    defined(
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: defined is a built in. Remove ()

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 11]
```diff
@@ -0,0 +1,1256 @@
...
+                          {$annotation_file}
+                    )
+                  )
+                {
+
+                    $name_prefix =
+                      $active_parameter_href->{snpsift_annotation_outinfo_key}
+                      {$annotation_file};
+                }
+                $info_key = $annotation_info_key;    #Database
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Move comment to line above and add whitespace after '#'

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 12]
```diff
@@ -0,0 +1,1256 @@
...
+            }
+
+            foreach my $contig ( @{$vcfparser_contigs_ref} ) {
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 13]
```diff
@@ -0,0 +1,1256 @@
...
+                          . $xargs_file_counter,
+                        verbosity => q{v},
+
+                    }
+                );
+                say {$XARGSFILEHANDLE} $NEWLINE;
+            }
+            $annotation_file_counter = $xargs_file_counter;
+        }
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 14]
```diff
@@ -0,0 +1,1256 @@
...
+            parameter_href => $parameter_href,
+            suffix_key     => q{variant_file_suffix},
+        }
+    );
+
+    my $vcfparser_analysis_type = $EMPTY_STR;
+
+    # Set default
+    my $vcfparser_contigs_ref = \@{ $file_info_href->{contigs_size_ordered} };
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 15]
```diff
@@ -0,0 +1,1256 @@
...
+                $info_key = $annotation_info_key;    #Database
+            }
+
+            foreach my $contig ( @{$vcfparser_contigs_ref} ) {
+
+                ## Get contig specific parameters
+                my $infile_path;
+
+                # First file per contig
+                if ( !$annotation_file_counter ) {
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: if( not $annotation_file_counter ) {

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 16]
```diff
@@ -0,0 +1,1256 @@
...
+                        config_file_path => catfile(
+                            $active_parameter_href->{snpeff_path},
+                            q{snpEff.config}
+                        ),
+                        database_path   => $annotation_file,
+                        FILEHANDLE      => $XARGSFILEHANDLE,
+                        infile_path     => $infile_path,
+                        info            => $info_key,
+                        name_prefix     => $name_prefix,
+                        stderrfile_path => $xargs_file_path_prefix
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Seem like we are writing both to 2> and 2>>. Should be able to remove 2> ie. stderrfile_path

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 17]
```diff
@@ -0,0 +1,1256 @@
...
+                    program_info_path  => $program_info_path,
+                    temp_directory     => $temp_directory,
+                    XARGSFILEHANDLE    => $XARGSFILEHANDLE,
+                    xargs_file_counter => $xargs_file_counter,
+                }
+            );
+
+            my $annotation_infile_number = $xargs_file_counter - 1;
+
+            foreach my $contig ( @{$vcfparser_contigs_ref} ) {
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 18]
```diff
@@ -0,0 +1,1256 @@
...
+    ## Unpack parameters
+    my $job_id_chain = $parameter_href->{$mip_program_name}{chain};
+
+    my ( $core_number, $time, $source_environment_cmd ) = get_module_parameters(
+        {
+            active_parameter_href => $active_parameter_href,
+            mip_program_name      => $mip_program_name,
+        }
+    );
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: $config_file_path = catfile(
+                            $active_parameter_href->{snpeff_path},
+                            q{snpEff.config}
+                        );

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 19]
```diff
@@ -0,0 +1,1256 @@
...
+                          $active_parameter_href->{snpsift_dbnsfp_file},
+                        FILEHANDLE  => $XARGSFILEHANDLE,
+                        infile_path => $file_path_prefix
+                          . $UNDERSCORE
+                          . $contig
+                          . $vcfparser_analysis_type
+                          . $infile_suffix
+                          . $DOT
+                          . $annotation_infile_number,
+                        stderrfile_path => $xargs_file_path_prefix
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Same here! I this by purpose? I cannot think of why this is added?

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 20]
```diff
@@ -0,0 +1,1256 @@
...
+                file_path          => $file_path,
+                program_info_path  => $program_info_path,
+                XARGSFILEHANDLE    => $XARGSFILEHANDLE,
+                xargs_file_counter => $xargs_file_counter,
+            }
+        );
+
+        my $annotation_infile_number = $xargs_file_counter - 1;
+
+        foreach my $contig ( @{$vcfparser_contigs_ref} ) {
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Add loop label

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 21]
```diff
@@ -0,0 +1,1256 @@
...
+                {
+                    FILEHANDLE  => $XARGSFILEHANDLE,
+                    infile_path => $file_path_prefix
+                      . $UNDERSCORE
+                      . $contig
+                      . $vcfparser_analysis_type
+                      . $infile_suffix
+                      . $DOT
+                      . $annotation_infile_number,
+                    stderrfile_path => $xargs_file_path_prefix
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: And here

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 22]
```diff
@@ -0,0 +1,1256 @@
...
+            $outfile_prefix
+          . $UNDERSCORE
+          . $file_info_href->{contigs_size_ordered}[0]
+          . $vcfparser_analysis_type
+          . $outfile_suffix;
+        add_program_outfile_to_sample_info(
+            {
+                sample_info_href => $sample_info_href,
+                program_name     => $program_name,
+                outdirectory     => $outfamily_directory,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Use path instead of outdirectory and outfile

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 23]
```diff
@@ -0,0 +1,1256 @@
...
+        ## Collect QC metadata info for later use
+        my $qc_snpeff_outfile =
+            $outfile_prefix
+          . $UNDERSCORE
+          . $file_info_href->{contigs_size_ordered}[0]
+          . $vcfparser_analysis_type
+          . $outfile_suffix;
+        add_program_outfile_to_sample_info(
+            {
+                sample_info_href => $sample_info_href,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Order AB

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 24]
```diff
@@ -0,0 +1,1256 @@
...
+                program_name     => $program_name,
+                outdirectory     => $outfamily_directory,
+                outfile          => $qc_snpeff_outfile,
+            }
+        );
+
+        slurm_submit_job_sample_id_dependency_add_to_family(
+            {
+                job_id_href             => $job_id_href,
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Order AB

File: lib/MIP/Recipes/Analysis/Snpeff.pm [hunk 25]
```diff
@@ -0,0 +1,1256 @@
...
+                infile_lane_prefix_href => $infile_lane_prefix_href,
+                sample_ids_ref   => \@{ $active_parameter_href->{sample_ids} },
+                family_id        => $family_id,
+                path             => $job_id_chain,
+                log              => $log,
+                sbatch_file_name => $file_path,
+            }
+        );
+
```
Comment (henrikstranneheim) [inline lib/MIP/Recipes/Analysis/Snpeff.pm]: Remove empty line

File: lib/MIP/Recipes/Analysis/Snpeff.pm (hunk 16 above)
Comment (northwestwitch) [reply lib/MIP/Recipes/Analysis/Snpeff.pm]: Or should I remove the stderrfile_path_append?

**Comment by henrikstranneheim** [comment]
👍

**Pull request merged**